### DISCUSSION TITLE
Question：workspace/executeCommand

### DISCUSSION THREAD

**Discussion by pl0301** [redhat-developer/lsp4ij#469] [Q&A]
I noticed that workspace/executeCommand is currently supported. Could you please tell me how to use it and where to input the commands I want to send to the server?

**Comment by angelozerr** [comment]
If command to execute is declared in code action, completion, please see https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#lsp-commands

If you need to execute a command in another location, you can use `CommandExecutor#executeCommand`

https://github.com/search?q=repo%3Aredhat-developer%2Flsp4ij%20CommandExecutor&type=code

**Comment by angelozerr** [comment]
@pl0301 is my answer is enough? Could we close this discussion or do you need more information or more feature?

**Comment by lstreckeisen** [comment]
I'm currently also trying to make use of `workspace/executeCommand`. From the documentation I gather that I should create an action class based on `LSPCommandAction` and register it with `<action ...` in the plugin.xml. But the `LSPCommandAction` never calls `commandPerformed`, since the action event doesn't have an LSPCommand in its event data.

As a workaround I created a regular action based on `AnAction` and executed the command via `CommandExecutor#executeCommand`. That works.
@angelozerr Is there anything else to do than what I described above, so that the LSPCommandAction is executed?

**Reply by angelozerr** [reply]
Could you share your code which doesn't work.

**Reply by angelozerr** [reply]
LSPCommandAction works with some LSP features like codelens, completion, inlay hint which return a command in the Json response. Is it your usecase?

**Reply by lstreckeisen** [reply]
No, my use case is to generate a PlantUML diagram based on the currently opened file.

I registered the action like this:
```
<action class="org.contextmapper.intellij.actions.PlantUMLAction"
                id="org.contextmapper.GeneratePlantUML"
                text="Generate PlantUML Diagrams"
                description="Generates plantUML diagrams"
        />
```

The action class looks like this (it's just a dummy so far):
```
class PlantUMLAction : LSPCommandAction() {
    override fun commandPerformed(
        p0: LSPCommand,
        p1: AnActionEvent
    ) {
        print("Test")
    }
}
```

**Reply by angelozerr** [reply]
LSPCommandAction must be used when language server returns a command to execute in codelens, inlay hint, etc

Your usecase seems just to consume workspace/executeCommand with a standard IntelliJ action, you must not use LSPCommandAction and use standard AnAction like you did.

**Reply by angelozerr** [reply]
See https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#execute-a-command

**Reply by lstreckeisen** [reply]
Ok thanks, good to know. That wasn't clear for me. 👍

**Reply by angelozerr** [reply]
If you think the doc is not clear, please create a PR to improve it. Thanks!

**Comment by angelozerr** [comment]
@lstreckeisen if you are using LSP4IJ in your plugin could you please:

 * create a PR in LSP4IJ to add it in https://github.com/redhat-developer/lsp4ij/tree/main?tab=readme-ov-file#who-is-using-lsp4ij
 * add a review in JetBrains Marketplace to help us to promote LSP4IJ.

Thanks!

**Reply by lstreckeisen** [reply]
Will do once my PoC was approved.
